// bucket-model/bucket-model.js
var _ = require('lodash');
var Decimal = require('decimal');

var sum = require('utils/sum');

/* @ngInject */
module.exports = function (BaseModel, BaseCollection, ContributionModel, UserModel, AuthService) {
  return BaseModel.extend({
    modelType: "Bucket",

    props: {
      name: "string",
      roundId: "number",
      description: "string",
      target: "decimal",
    },

    children: {
      user: UserModel,
    },

    collections: {
      contributions: BaseCollection.extend({
        model: ContributionModel,
      }),
    },

    initialize: function () {
      // listen to contributions events
      this.listenTo(this.contributions, 'all',
        this._getEventBubblingHandler(this.contributions)
      );

      this.listenTo(this.contributions, 'add', function (contribution) {
        // attach bucketId to contributions, as the data
        // returned from the API does not have this
        contribution.bucketId = this.id;
      });
    },

    derived: {
      contributionCount: {
        deps: ['contributions'],
        cache: false,
        fn: function () {
          return this.contributions.filter(function (contribution) {
            return contribution.amount.gt(0);
          }).length;
        },
      },
      contributionTotal: {
        deps: ['contributions'],
        cache: false,
        fn: function () {
          return sum(this.contributions.pluck("amount"));
        },
      },
      myContributionIndex: {
        deps: ['contributions'],
        cache: false,
        fn: function () {
          return this.contributions.findIndex(function (contribution) {
            return contribution.user.id === AuthService.getCurrentUser().id;
          });
        },
      },
      myContribution: {
        deps: ['myContributionIndex', 'contributions'],
        cache: false,
        fn: function () {
          var index = this.myContributionIndex;
          if (index !== -1) {
            return this.contributions.models[index];
          } else {
            var currentUser = AuthService.getCurrentUser();
            return new ContributionModel({
              user: {
                id: currentUser.id,
                name: currentUser.name,
              },
              bucketId: this.id,
              amount: 0
            });
          }
        },
      },
      groupContributions: {
        deps: ['contributions'],
        cache: false,
        fn: function () {
          return _.filter(contributions, function (contribution) {
            return contribution.user.id !== AuthService.getCurrentUser.id;
          });
        },
      },
      percentageFunded: {
        deps: ['contributionTotal', 'target'],
        cache: false,
        fn: function () {
          if (!this.target.gt(0)) {
            return new Decimal(0);
          }
          return this.contributionTotal.div(this.target);
        },
      },
      myContributionPercentage: {
        deps: ['myContribution', 'target'],
        cache: false,
        fn: function () {
          if (!this.target.gt(0)) {
            return new Decimal(0);
          }
          return this.myContribution.amount.div(this.target);
        },
      },
      groupContributionPercentage: {
        deps: ['percentageFunded', 'myContributionPercentage'],
        cache: false,
        fn: function () {
          return this.percentageFunded.minus(this.myContributionPercentage);
        },
      },
      groupContribution: {
        deps: ['contributionTotal', 'myContribution'],
        cache: false,
        fn: function () {
          return this.contributionTotal.minus(this.myContribution.amount);
        },
      },
    },

    serialize: function () {
      var srlzd = BaseModel.prototype.serialize.call(this);
      srlzd.userId = srlzd.user.id;
      delete srlzd.user;
      return srlzd;
    },

    getContributionsByUser: function () {
      var contributions = {};
      this.contributions.forEach(function (contribution) {
        contributions[contribution.user.id] = contribution;
      });
      return contributions;
    },
  });
};
